### PR TITLE
Use parse_url kernel for HOST parsing

### PR DESCRIPTION
**Pull request by thirtiseven**
Contributes to https://github.com/NVIDIA/spark-rapids/issues/8963

This PR add plugin support for HOST parsing in parse_url.

Depends on JNI PR: https://github.com/NVIDIA/spark-rapids-jni/pull/1569

Depends on: #9481

### CODE REVIEW

File: integration_tests/src/main/python/url_test.py [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+    "http://[2001:db8:3:4::192.0.2.33]",
+    "http://[64:ff9b::192.0.2.33]"
+]
+
+edge_cases_gen = SetValuesGen(StringType(), edge_cases)
+
+url_gen = StringGen(url_pattern)
+    
+@pytest.mark.parametrize('data_gen', [url_gen, edge_cases_gen], ids=idfn)
+def test_parse_url_protocol(data_gen):
```
Comment (winningsix) [inline integration_tests/src/main/python/url_test.py:150]: nit: necessary to have some negative cases for protocol and host? To validate same exception behavior as Spark. Not sure the expected behavior for NaN and null string literal.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/urlFunctions.scala [hunk 2]
```diff
@@ -0,0 +1,108 @@
+/*
```
Comment (winningsix) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/urlFunctions.scala:1]: Make this file name capitalized.

File: integration_tests/src/main/python/url_test.py [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+                "a",
+                "parse_url(a, 'PROTOCOL')"
+                ))
+    
+@pytest.mark.parametrize('data_gen', [url_gen, edge_cases_gen], ids=idfn)
+def test_parse_url_host(data_gen):
+    assert_gpu_and_cpu_are_equal_collect(
+            lambda spark : unary_op_df(spark, data_gen).selectExpr(
+                "a",
+                "parse_url(a, 'HOST')"
```
Comment (winningsix) [inline integration_tests/src/main/python/url_test.py:162]: nit: make ```HOST``` parameterized. Same as you did for ```unsupported_part```.

File: integration_tests/src/main/python/url_test.py [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+                ))
+    
+unsupported_part = ['HOST', 'PATH', 'QUERY', 'REF', 'FILE', 'AUTHORITY', 'USERINFO']
```
Comment (winningsix) [inline integration_tests/src/main/python/url_test.py:165]: nit: typo ```unsupported_parts```

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala [hunk 5]
```diff
@@ -3227,6 +3227,34 @@ object GpuOverrides extends Logging {
...
+          // Should really be an OptionalParam
+          Some(RepeatingParamCheck("key", TypeSig.lit(TypeEnum.STRING), TypeSig.STRING))),
+      (a, conf, p, r) => new ExprMeta[ParseUrl](a, conf, p, r) {
+        override def tagExprForGpu(): Unit = {
+          if (a.failOnError) {
+            willNotWorkOnGpu("Fail on error is not supported on GPU when parsing urls.")
+          }
+
+          extractStringLit(a.children(1)).map(_.toUpperCase) match {
+            case Some(GpuParseUrl.PROTOCOL) | Some(GpuParseUrl.HOST) =>
```
Comment (winningsix) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala:3245]: nit: Better to move this into  ```GpuParseUrl``` as a util method as a list for supported URL pattern.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/urlFunctions.scala [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+  def doColumnar(url: GpuColumnVector, partToExtract: GpuScalar): ColumnVector = {
+    val part = partToExtract.getValue.asInstanceOf[UTF8String].toString
+    part match {
+      case PROTOCOL =>
+        ParseURI.parseURIProtocol(url.getBase)
+      case HOST =>
+        ParseURI.parseURIHost(url.getBase)
+      case PATH | QUERY | REF | FILE | AUTHORITY | USERINFO =>
+        throw new UnsupportedOperationException(s"$this is not supported partToExtract=$part")
```
Comment (winningsix) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/urlFunctions.scala:60]: Nit: can we also mention what URL type we supported in err msg?

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/urlFunctions.scala [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+
+  override def columnarEval(batch: ColumnarBatch): GpuColumnVector = {
+    if (children.size == 2) {
+      val Seq(url, partToExtract) = children
+      withResourceIfAllowed(url.columnarEval(batch)) { urls =>
+        withResourceIfAllowed(partToExtract.columnarEvalAny(batch)) { parts =>
+          parts match {
+            case partScalar: GpuScalar =>
+              GpuColumnVector.from(doColumnar(urls, partScalar), dataType)
+            case _ =>
```
Comment (winningsix) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/urlFunctions.scala:83]: Just curious in which case we will hit this exception.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/urlFunctions.scala [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+      val Seq(url, partToExtract, key) = children
+      withResourceIfAllowed(url.columnarEval(batch)) { urls =>
+        withResourceIfAllowed(partToExtract.columnarEvalAny(batch)) { parts =>
+          withResourceIfAllowed(key.columnarEvalAny(batch)) { keys =>
+            (urls, parts, keys) match {
+              case (urlCv: GpuColumnVector, partScalar: GpuScalar, keyScalar: GpuScalar) =>
+                GpuColumnVector.from(doColumnar(urlCv, partScalar, keyScalar), dataType)
+              case _ =>
+                throw new 
+                    UnsupportedOperationException(s"Cannot columnar evaluate expression: $this")
```
Comment (winningsix) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/urlFunctions.scala:101]: ditto.

File: tests/src/test/scala/com/nvidia/spark/rapids/UrlFunctionsSuite.scala [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+    // [In search of the perfect URL validation regex](https://mathiasbynens.be/demo/url-regex)
+    Seq[String](
+      "http://foo.com/blah_blah",
+      "http://foo.com/blah_blah/",
+      "http://foo.com/blah_blah_(wikipedia)",
+      "http://foo.com/blah_blah_(wikipedia)_(again)",
+      "http://www.example.com/wpstyle/?p=364",
+      "https://www.example.com/foo/?bar=baz&inga=42&quux",
+      "http://✪df.ws/123",
+      // "http://userid:password@example.com:8080",
```
Comment (winningsix) [inline tests/src/test/scala/com/nvidia/spark/rapids/UrlFunctionsSuite.scala:33]: Remove these if we don't want.

File: tests/src/test/scala/com/nvidia/spark/rapids/UrlFunctionsSuite.scala [hunk 10]
```diff
@@ -0,0 +1,198 @@
+/*
```
Comment (winningsix) [inline tests/src/test/scala/com/nvidia/spark/rapids/UrlFunctionsSuite.scala:1]: Seems duplicated to ```url_test.py```. We can just keep ```url_test.py``` which can be run in parallel.

File: integration_tests/src/main/python/url_test.py (hunk 1 above)
Comment (thirtiseven) [reply integration_tests/src/main/python/url_test.py:150]: Yes, we have many negative cases that spark returns null in tests, and we will also test against a large dataset that includes many negative cases.

File: tests/src/test/scala/com/nvidia/spark/rapids/UrlFunctionsSuite.scala (hunk 10 above)
Comment (thirtiseven) [reply tests/src/test/scala/com/nvidia/spark/rapids/UrlFunctionsSuite.scala:1]: Yes, it is just make it easy to test for now. Will remove it before it's ready for review.

File: tests/src/test/scala/com/nvidia/spark/rapids/UrlFunctionsSuite.scala (hunk 9 above)
Comment (thirtiseven) [reply tests/src/test/scala/com/nvidia/spark/rapids/UrlFunctionsSuite.scala:33]: They are failing tests now, will uncomment them after the kernel fixes these cases.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala (hunk 5 above)
Comment (thirtiseven) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala]: done.

File: tests/src/test/scala/com/nvidia/spark/rapids/UrlFunctionsSuite.scala (hunk 10 above)
Comment (thirtiseven) [reply tests/src/test/scala/com/nvidia/spark/rapids/UrlFunctionsSuite.scala]: Removed.

**Comment by thirtiseven** [comment]
build

File: integration_tests/src/main/python/url_test.py [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+unsupported_parts = ['PATH', 'QUERY', 'REF', 'FILE', 'AUTHORITY', 'USERINFO']
+    
+@pytest.mark.parametrize('data_gen', [url_gen, edge_cases_gen], ids=idfn)
+@pytest.mark.parametrize('part', supported_parts, ids=idfn)
+def test_parse_url_supported(data_gen, part):
+    assert_gpu_and_cpu_are_equal_collect(
+            lambda spark : unary_op_df(spark, data_gen).selectExpr(
+                "a",
+                "parse_url(a, '" + part + "')"
+                ))
```
Comment (ttnghia) [inline integration_tests/src/main/python/url_test.py:159]: Probably the code needs to be reformated: The indentations are incorrect.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuParseUrl.scala [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+    if (children.size == 2) {
+      val Seq(url, partToExtract) = children
+      withResourceIfAllowed(url.columnarEval(batch)) { urls =>
+        withResourceIfAllowed(partToExtract.columnarEvalAny(batch)) { parts =>
+          parts match {
+            case partScalar: GpuScalar =>
+              GpuColumnVector.from(doColumnar(urls, partScalar), dataType)
+            case _ =>
+              throw new 
+                  UnsupportedOperationException(s"Cannot columnar evaluate expression: $this")
```
Comment (ttnghia) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuParseUrl.scala:95]: Reformat code?

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuParseUrl.scala [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+    }
+  }
+
+  def doColumnar(url: GpuColumnVector, partToExtract: GpuScalar, key: GpuScalar): ColumnVector = {
+    val part = partToExtract.getValue.asInstanceOf[UTF8String].toString
+    if (part != QUERY) {
+      // return a null columnvector
+      return ColumnVector.fromStrings(null, null)
+    }
+    throw new UnsupportedOperationException(s"$this only supports partToExtract = PROTOCOL")
```
Comment (ttnghia) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuParseUrl.scala:82]: The message seems to be vague here. We say we are supporting PROTOCOL and HOST so here needs more explanation.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuParseUrl.scala [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+              GpuColumnVector.from(doColumnar(urls, partScalar), dataType)
+            case _ =>
+              throw new 
+                  UnsupportedOperationException(s"Cannot columnar evaluate expression: $this")
+          }
+        }
+      }
+    } else {
+      // 3-arg, i.e. QUERY with key
+      assert(children.size == 3)
```
Comment (ttnghia) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuParseUrl.scala:101]: How about throwing `IllegalArgument` on object constructor (`object GpuParseUrl#apply()`)?
```

```

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuParseUrl.scala (hunk 12 above)
Comment (thirtiseven) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuParseUrl.scala]: done.

File: integration_tests/src/main/python/url_test.py (hunk 11 above)
Comment (thirtiseven) [reply integration_tests/src/main/python/url_test.py]: done.

**Comment by thirtiseven** [comment]
dependent PROTOCOL PR is merged, so this one is ready to merge too.

**Review by revans2 — APPROVED** [review]
Looks good

**Comment by revans2** [comment]
build
